Sources/Plasma/FeatureLib/pfPython/plPythonCallable.h

```diff
@@ -54,102 +55,90 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
+                size_t i = 1; // Don't touch the first slot...
+                ((argsObjs[i++] = ConvertFrom(std::forward<Args>(args))), ...);
+            }
 
-    inline void IBuildTupleArg(PyObject* tuple, size_t idx, uint16_t value)
-    {
-        PyTuple_SET_ITEM(tuple, idx, PyLong_FromSize_t(value));
-    }
+            // Now, we perform the Python call with the resulting C++ compile-time magic.
+            plProfile_BeginTiming(PythonUpdate);
+            pyObjectRef result = _PyObject_Vectorcall(
+                callable,
+                &argsObjs[1], // This is the optimization mentioned earlier.
+                (_NArgsfT - 1) | PY_VECTORCALL_ARGUMENTS_OFFSET, // Indicates we are passing args[1] instead of args[0]
+                nullptr
+            );
+            plProfile_EndTiming(PythonUpdate);
 
-    inline void IBuildTupleArg(PyObject* tuple, size_t idx, uint32_t value)
-    {
-        PyTuple_SET_ITEM(tuple, idx, PyLong_FromSize_t(value));
-    }
+            for (size_t i = 1; i < _NArgsfT; ++i)
+                Py_XDECREF(argsObjs[i]);
+            return result;
+        }
+    };
 
-    inline void IBuildTupleArg(PyObject* tuple, size_t idx, wchar_t value)
+    template<typename... Args>
+    inline pyObjectRef CallObject(const pyObjectRef& callable, Args&&... args)
     {
-        PyTuple_SET_ITEM(tuple, idx, PyUnicode_FromFormat("%c", (int)value));
+        return CallObject(callable.Get(), std::forward<Args>(args)...);
     }
 
-    template<size_t Size, typename Arg>
-    inline void BuildTupleArgs(PyObject* tuple, Arg&& arg)
+    template<typename... Args>
+    inline pyObjectRef CallObject(PyObject* callable, Args&&... args)
     {
-        IBuildTupleArg(tuple, (Size - 1), std::forward<Arg>(arg));
-    }
+        hsAssert(PyCallable_Check(callable), "Trying to call a non-callable, eh?");
 
-    template<size_t Size, typename Arg0, typename... Args>
-    inline void BuildTupleArgs(PyObject* tuple, Arg0&& arg0, Args&&... args)
-    {
-        IBuildTupleArg(tuple, (Size - (sizeof...(args) + 1)), std::forward<Arg0>(arg0));
-        BuildTupleArgs<Size>(tuple, std::forward<Args>(args)...);
+        // The point of all this is to use Python's new "vectorcall" calling convention.
+        // PSF claims that is is faster -- the most evident improvement is that we don't
+        // have to allocate a transitional tuple object to pass the arguments.
+        if constexpr (sizeof...(args) == 1) {
+            // Use Python's built-in vectorcall optimization for one argument.
+            pyObjectRef arg = ConvertFrom(std::forward<Args>(args)...);
+            plProfile_BeginTiming(PythonUpdate);
+            pyObjectRef result = _PyObject_CallOneArg(callable, arg.Get());
+            plProfile_EndTiming(PythonUpdate);
+            return result;
+        } else if constexpr (sizeof...(args) == 0) {
+            plProfile_BeginTiming(PythonUpdate);
+#if PY_VERSION_HEX >= 0x03090000
```

Comment:
I'm not familiar with the hex comparison here. Is this 3.9 or greater, so you can use the built-in versus the provisional vectorcall methods?

Reply:
IIRC the zero argument specialization for vectorcall was added in Python 3.9, so I had to add manual handling for 3.8.